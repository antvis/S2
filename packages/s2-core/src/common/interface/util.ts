export type DeepRequired<T extends Record<string, any>> = {
  [K in keyof T]-?: NonNullable<T[K]> extends Record<string, any>
    ? DeepRequired<NonNullable<T[K]>>
    : NonNullable<T[K]>;
};
