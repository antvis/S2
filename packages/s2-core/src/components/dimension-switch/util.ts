import { DimensionType, DimensionItem } from './dimension';

export interface OperatedType {
  [type: string]: string[];
}

export const getDimensionsByPredicate = (
  data: DimensionType[],
  predicate: (value: DimensionItem) => boolean,
) => {
  return data.reduce((obj, t) => {
    obj[t.type] = t.items.filter(predicate).map((i) => i.id);
    return obj;
  }, {} as OperatedType);
};
