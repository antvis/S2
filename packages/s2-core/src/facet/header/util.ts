import { isNil } from 'lodash';

export const getCellPadding = () => {
  const padding = [12, 4, 12, 4];
  const left = isNil(padding[3]) ? 0 : padding[3];
  const right = isNil(padding[1]) ? 0 : padding[1];
  const top = isNil(padding[0]) ? 0 : padding[0];
  const bottom = isNil(padding[2]) ? 0 : padding[2];
  return {
    left,
    right,
    top,
    bottom,
  };
};
