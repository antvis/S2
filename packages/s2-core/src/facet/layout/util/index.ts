export { default as getDimsConditionByNode } from './get-dims-condition-by-node';
export { default as getColHeight } from './get-col-height';
export { default as getColWidth } from './get-col-width';
export { default as getRowWidth } from './get-row-width';
export { default as processColsList } from './process-cols-list';
export { default as getHeaderHierarchyQuery } from './get-header-hierarchy-query';
