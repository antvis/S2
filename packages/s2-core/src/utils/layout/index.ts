export { generateId } from './generate-id';
export * from './transform-custom-tree-items';
export * from './frozen';
