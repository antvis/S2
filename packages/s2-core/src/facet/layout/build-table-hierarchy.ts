import { map } from 'lodash';
import { TableHeaderParams } from '@/facet/layout/interface';
import { SERIES_NUMBER_FIELD } from '@/common/constant';
import { i18n } from '@/common/i18n';
import { generateHeaderNodes } from '@/utils/layout/generate-header-nodes';

export const buildTableHierarchy = (params: TableHeaderParams) => {
  const { facetCfg, hierarchy, parentNode } = params;
  const { columns, spreadsheet, dataSet } = facetCfg;

  const hiddenColumnsDetail = spreadsheet.store.get('hiddenColumnsDetail');
  const showSeriesNumber = facetCfg?.showSeriesNumber;

  const displayedColumns = columns.filter((column) => {
    if (!hiddenColumnsDetail) {
      return true;
    }
    return hiddenColumnsDetail.every((detail) => {
      return detail.hideColumnNodes.every((node) => {
        return node.field !== column;
      });
    });
  });
  const fields = [...displayedColumns];

  const fieldValues = map(displayedColumns, (val) => dataSet.getFieldName(val));

  if (showSeriesNumber) {
    fields.unshift(SERIES_NUMBER_FIELD);
    fieldValues.unshift(i18n('序号'));
  }

  generateHeaderNodes({
    currentField: displayedColumns[0],
    fields,
    fieldValues,
    facetCfg,
    hierarchy,
    parentNode,
    level: 0,
    query: {},
    addMeasureInTotalQuery: false,
    addTotalMeasureInTotal: false,
  });
};
