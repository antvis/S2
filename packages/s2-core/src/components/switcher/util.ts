import { filter, flatten, isEmpty, isNil, map } from 'lodash';
import { DraggableLocation } from 'react-beautiful-dnd';
import {
  FieldType,
  MAX_DIMENSION_COUNT,
  SWITCHER_PREFIX_CLS,
} from './constant';
import { SwitcherItem, SwitchResult, SwitchState } from './interface';
import { getClassNameWithPrefix } from '@/utils/get-classnames';

export const getSwitcherClassName = (...classNames: string[]) =>
  getClassNameWithPrefix(SWITCHER_PREFIX_CLS, ...classNames);

export const getNonEmptyFieldCount = (state: SwitchState) => {
  return [
    state[FieldType.Rows],
    state[FieldType.Cols],
    state[FieldType.Values],
  ].reduce((sum, value) => sum + (isEmpty(value) ? 0 : 1), 0);
};

export const getMainLayoutClassName = (nonEmptyCount: number) => {
  switch (nonEmptyCount) {
    case 1:
      return getSwitcherClassName('content', 'one-dimension');
    case 2:
      return getSwitcherClassName('content', 'two-dimensions');
    default:
      return getSwitcherClassName('content', 'three-dimensions');
  }
};

export const showDimensionCrossRows = (nonEmptyCount: number) =>
  nonEmptyCount < MAX_DIMENSION_COUNT;

export const isMeasureType = (fieldType: FieldType) =>
  fieldType === FieldType.Values;

export const moveItem = (
  source: SwitcherItem[],
  destination: SwitcherItem[],
  droppableSource: DraggableLocation,
  droppableDestination: DraggableLocation,
): SwitchState => {
  // change order in same column
  if (droppableDestination.droppableId === droppableSource.droppableId) {
    const updatingDestination = [...destination];
    const [removed] = updatingDestination.splice(droppableSource.index, 1);
    updatingDestination.splice(droppableDestination.index, 0, removed);
    return {
      [droppableDestination.droppableId]: updatingDestination,
    };
  }
  // move to other column
  const updatingSource = [...source];
  const updatingDestination = [...destination];

  const [removed] = updatingSource.splice(droppableSource.index, 1);
  updatingDestination.splice(droppableDestination.index, 0, removed);

  return {
    [droppableSource.droppableId]: updatingSource,
    [droppableDestination.droppableId]: updatingDestination,
  };
};

export const checkItem = (
  source: SwitcherItem[],
  checked: boolean,
  id: string,
  parentId?: string,
): SwitcherItem[] => {
  const target: SwitcherItem = {
    ...source.find((item) => item.id === (parentId ?? id)),
  };

  // 有 parentId 时，说明是第二层次项的改变
  if (parentId) {
    target.children = map(target.children, (item) => ({
      ...item,
      checked: item.id === id ? checked : item.checked,
    }));
  } else {
    target.checked = checked;
    target.children = map(target.children, (item) => ({
      ...item,
      checked: checked,
    }));
  }

  return source.map((item) => (item.id === target.id ? target : item));
};

export const generateSwitchResult = (state: SwitchState): SwitchResult => {
  // rows and cols can't be hidden
  const rows = map(state[FieldType.Rows], 'id');
  const cols = map(state[FieldType.Cols], 'id');

  // flatten all values and derived values
  const values = flatten(
    map(state[FieldType.Values], (item) => {
      const children = map(item.children, 'id');
      return [item.id, ...children];
    }),
  );

  const filterHiddenValues = (item: SwitcherItem) => item.checked === false;

  //  get all hidden values
  const hiddenValues = flatten(
    map(filter(state[FieldType.Values], filterHiddenValues), (item) => {
      const hiddenChildren = map(
        filter(item.children, filterHiddenValues),
        'id',
      );
      return [item.id, ...hiddenChildren];
    }),
  );

  return { rows, cols, values, hiddenValues };
};
