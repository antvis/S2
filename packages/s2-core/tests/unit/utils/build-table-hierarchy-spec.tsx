import { act } from 'react-dom/test-utils';
import 'antd/dist/antd.min.css';
import {
  S2DataConfig,
  S2Options,
  SheetComponent,
  SpreadSheet,
  Node,
  Hierarchy,
} from '../../../src';
import { buildTableHierarchy } from '../../../src/facet/layout/build-table-hierarchy';
import { getContainer, getMockData } from '../../util/helpers';
import ReactDOM from 'react-dom';
import React from 'react';

const data = getMockData('../data/tableau-supermarket.csv');

let spreadsheetIns: SpreadSheet;

const getSpreadSheet = (
  dom: string | HTMLElement,
  dataCfg: S2DataConfig,
  options: S2Options,
) => {
  spreadsheetIns = new SpreadSheet(dom, dataCfg, options);
  return spreadsheetIns;
};

const getDataCfg = () => {
  return {
    fields: {
      columns: [
        'area',
        'province',
        'city',
        'type',
        'sub_type',
        'profit',
        'count',
      ],
    },
    data,
  };
};

const getOptions = () => {
  return {
    width: 800,
    height: 600,
    showSeriesNumber: true,
    mode: 'table',
    style: {
      colCfg: {
        colWidthType: 'compact',
      },
      cellCfg: {
        height: 32,
      },
      device: 'pc',
    },
  };
};

const getTheme = () => {
  return {};
};

function MainLayout(props) {
  const [options, setOptions] = React.useState(props.options);
  const [dataCfg, setDataCfg] = React.useState(props.dataCfg);

  return (
    <div>
      <div style={{ display: 'inline-block' }}></div>
      <SheetComponent
        dataCfg={dataCfg}
        adaptive={false}
        options={options}
        theme={props.theme}
        spreadsheet={getSpreadSheet}
      />
    </div>
  );
}

describe('buildTableHierarchy', () => {
  test('sort action with number arr', () => {
    act(() => {
      ReactDOM.render(
        <MainLayout
          dataCfg={getDataCfg()}
          options={getOptions()}
          theme={getTheme()}
        />,
        getContainer(),
      );
    });

    const rootNode = Node.rootNode();
    const hierarchy = new Hierarchy();

    buildTableHierarchy({
      parentNode: rootNode,
      facetCfg: spreadsheetIns.facet.cfg,
      hierarchy,
    });

    expect(JSON.stringify(rootNode)).toMatchSnapshot();
    expect(JSON.stringify(hierarchy)).toMatchSnapshot();
  });
});
