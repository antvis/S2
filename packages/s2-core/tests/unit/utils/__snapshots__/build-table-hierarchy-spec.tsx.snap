// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildTableHierarchy should generate nodes and hierarchy correctly 1`] = `"{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"children\\":[{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]序号\\",\\"key\\":\\"$$series_number$$\\",\\"value\\":\\"序号\\",\\"label\\":\\"序号\\",\\"rowIndex\\":0,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"$$series_number$$\\",\\"query\\":{\\"$$series_number$$\\":\\"序号\\"},\\"isTotalMeasure\\":false},{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]area\\",\\"key\\":\\"area\\",\\"value\\":\\"area\\",\\"label\\":\\"area\\",\\"rowIndex\\":1,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"area\\",\\"query\\":{\\"area\\":\\"area\\"},\\"isTotalMeasure\\":false},{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]province\\",\\"key\\":\\"province\\",\\"value\\":\\"province\\",\\"label\\":\\"province\\",\\"rowIndex\\":2,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"province\\",\\"query\\":{\\"province\\":\\"province\\"},\\"isTotalMeasure\\":false},{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]city\\",\\"key\\":\\"city\\",\\"value\\":\\"city\\",\\"label\\":\\"city\\",\\"rowIndex\\":3,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"city\\",\\"query\\":{\\"city\\":\\"city\\"},\\"isTotalMeasure\\":false},{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]type\\",\\"key\\":\\"type\\",\\"value\\":\\"type\\",\\"label\\":\\"type\\",\\"rowIndex\\":4,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"type\\",\\"query\\":{\\"type\\":\\"type\\"},\\"isTotalMeasure\\":false},{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]sub_type\\",\\"key\\":\\"sub_type\\",\\"value\\":\\"sub_type\\",\\"label\\":\\"sub_type\\",\\"rowIndex\\":5,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"sub_type\\",\\"query\\":{\\"sub_type\\":\\"sub_type\\"},\\"isTotalMeasure\\":false},{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]profit\\",\\"key\\":\\"profit\\",\\"value\\":\\"profit\\",\\"label\\":\\"profit\\",\\"rowIndex\\":6,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"profit\\",\\"query\\":{\\"profit\\":\\"profit\\"},\\"isTotalMeasure\\":false},{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]count\\",\\"key\\":\\"count\\",\\"value\\":\\"count\\",\\"label\\":\\"count\\",\\"rowIndex\\":7,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"count\\",\\"query\\":{\\"count\\":\\"count\\"},\\"isTotalMeasure\\":false}],\\"padding\\":0,\\"id\\":\\"root\\",\\"key\\":\\"\\",\\"value\\":\\"\\",\\"label\\":\\"\\"}"`;

exports[`buildTableHierarchy should generate nodes and hierarchy correctly 2`] = `"{\\"width\\":0,\\"height\\":0,\\"maxLevel\\":0,\\"sampleNodesForAllLevels\\":[{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]序号\\",\\"key\\":\\"$$series_number$$\\",\\"value\\":\\"序号\\",\\"label\\":\\"序号\\",\\"rowIndex\\":0,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"$$series_number$$\\",\\"query\\":{\\"$$series_number$$\\":\\"序号\\"},\\"isTotalMeasure\\":false}],\\"sampleNodeForLastLevel\\":{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]序号\\",\\"key\\":\\"$$series_number$$\\",\\"value\\":\\"序号\\",\\"label\\":\\"序号\\",\\"rowIndex\\":0,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"$$series_number$$\\",\\"query\\":{\\"$$series_number$$\\":\\"序号\\"},\\"isTotalMeasure\\":false},\\"allNodesWithoutRoot\\":[{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]序号\\",\\"key\\":\\"$$series_number$$\\",\\"value\\":\\"序号\\",\\"label\\":\\"序号\\",\\"rowIndex\\":0,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"$$series_number$$\\",\\"query\\":{\\"$$series_number$$\\":\\"序号\\"},\\"isTotalMeasure\\":false},{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]area\\",\\"key\\":\\"area\\",\\"value\\":\\"area\\",\\"label\\":\\"area\\",\\"rowIndex\\":1,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"area\\",\\"query\\":{\\"area\\":\\"area\\"},\\"isTotalMeasure\\":false},{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]province\\",\\"key\\":\\"province\\",\\"value\\":\\"province\\",\\"label\\":\\"province\\",\\"rowIndex\\":2,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"province\\",\\"query\\":{\\"province\\":\\"province\\"},\\"isTotalMeasure\\":false},{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]city\\",\\"key\\":\\"city\\",\\"value\\":\\"city\\",\\"label\\":\\"city\\",\\"rowIndex\\":3,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"city\\",\\"query\\":{\\"city\\":\\"city\\"},\\"isTotalMeasure\\":false},{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]type\\",\\"key\\":\\"type\\",\\"value\\":\\"type\\",\\"label\\":\\"type\\",\\"rowIndex\\":4,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"type\\",\\"query\\":{\\"type\\":\\"type\\"},\\"isTotalMeasure\\":false},{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]sub_type\\",\\"key\\":\\"sub_type\\",\\"value\\":\\"sub_type\\",\\"label\\":\\"sub_type\\",\\"rowIndex\\":5,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"sub_type\\",\\"query\\":{\\"sub_type\\":\\"sub_type\\"},\\"isTotalMeasure\\":false},{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]profit\\",\\"key\\":\\"profit\\",\\"value\\":\\"profit\\",\\"label\\":\\"profit\\",\\"rowIndex\\":6,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"profit\\",\\"query\\":{\\"profit\\":\\"profit\\"},\\"isTotalMeasure\\":false},{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]count\\",\\"key\\":\\"count\\",\\"value\\":\\"count\\",\\"label\\":\\"count\\",\\"rowIndex\\":7,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"count\\",\\"query\\":{\\"count\\":\\"count\\"},\\"isTotalMeasure\\":false}],\\"indexNode\\":[{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]序号\\",\\"key\\":\\"$$series_number$$\\",\\"value\\":\\"序号\\",\\"label\\":\\"序号\\",\\"rowIndex\\":0,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"$$series_number$$\\",\\"query\\":{\\"$$series_number$$\\":\\"序号\\"},\\"isTotalMeasure\\":false},{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]area\\",\\"key\\":\\"area\\",\\"value\\":\\"area\\",\\"label\\":\\"area\\",\\"rowIndex\\":1,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"area\\",\\"query\\":{\\"area\\":\\"area\\"},\\"isTotalMeasure\\":false},{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]province\\",\\"key\\":\\"province\\",\\"value\\":\\"province\\",\\"label\\":\\"province\\",\\"rowIndex\\":2,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"province\\",\\"query\\":{\\"province\\":\\"province\\"},\\"isTotalMeasure\\":false},{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]city\\",\\"key\\":\\"city\\",\\"value\\":\\"city\\",\\"label\\":\\"city\\",\\"rowIndex\\":3,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"city\\",\\"query\\":{\\"city\\":\\"city\\"},\\"isTotalMeasure\\":false},{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]type\\",\\"key\\":\\"type\\",\\"value\\":\\"type\\",\\"label\\":\\"type\\",\\"rowIndex\\":4,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"type\\",\\"query\\":{\\"type\\":\\"type\\"},\\"isTotalMeasure\\":false},{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]sub_type\\",\\"key\\":\\"sub_type\\",\\"value\\":\\"sub_type\\",\\"label\\":\\"sub_type\\",\\"rowIndex\\":5,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"sub_type\\",\\"query\\":{\\"sub_type\\":\\"sub_type\\"},\\"isTotalMeasure\\":false},{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]profit\\",\\"key\\":\\"profit\\",\\"value\\":\\"profit\\",\\"label\\":\\"profit\\",\\"rowIndex\\":6,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"profit\\",\\"query\\":{\\"profit\\":\\"profit\\"},\\"isTotalMeasure\\":false},{\\"x\\":0,\\"y\\":0,\\"width\\":0,\\"height\\":0,\\"colIndex\\":-1,\\"level\\":0,\\"isLeaf\\":true,\\"children\\":[],\\"padding\\":0,\\"id\\":\\"root[&]count\\",\\"key\\":\\"count\\",\\"value\\":\\"count\\",\\"label\\":\\"count\\",\\"rowIndex\\":7,\\"isTotals\\":false,\\"isCollapsed\\":false,\\"field\\":\\"count\\",\\"query\\":{\\"count\\":\\"count\\"},\\"isTotalMeasure\\":false}]}"`;
